iface/get.go

```diff
@@ -2,79 +2,159 @@ package iface
 
 import (
 	"fmt"
-	"go/token"
+	"go/ast"
 	"go/types"
+	"path/filepath"
 	"sort"
 	"strings"
 
-	"github.com/pkg/errors"
 	"golang.org/x/tools/go/packages"
 )
 
-func GetInterface(dir, ifaceName string) (Interface, error) {
-	cfg := &packages.Config{Mode: packages.LoadSyntax}
-	pkgs, err := packages.Load(cfg, dir)
-	if err != nil {
-		return Interface{}, errors.Wrap(err, "error loading package info")
-	}
+// ifaceInfo represents syntactic and type information associated with an
+// interface definition.
+type ifaceInfo struct {
+	pkg        *packages.Package
+	fileNode   *ast.File
+	object     types.Object
+	outputPath string
+}
+
+// GetAllInterfaces searches the given packages for interfaces annotated with a
+// "go:mock" directive, returning text-template-friendly representations.
+func GetAllInterfaces(pkgs []*packages.Package) ([]Interface, error) {
+	var allInfo []ifaceInfo
+	for _, pkg := range pkgs {
+		for _, fileNode := range pkg.Syntax {
+			ast.Inspect(fileNode, func(node ast.Node) bool {
+				// Only consider declarations with godoc comments.
+				decl, isGenDecl := node.(*ast.GenDecl)
+				if !isGenDecl || decl.Doc == nil {
+					return true
+				}
+				// Look for the first go:mock directive in the comments.
+				for _, comment := range decl.Doc.List {
+					args := strings.TrimPrefix(comment.Text, "//go:mock")
+					if args == comment.Text {
+						continue
+					}
+
+					// Build a qualified output pathname based on the output
+					// filename (or a default) and the input filepath.
+					var (
+						inputPath  = pkg.Fset.File(fileNode.Pos()).Name()
+						outputFile = strings.TrimSpace(args)
+						outputPath string
+					)
+					if outputFile == "" {
+						outputPath = strings.TrimSuffix(inputPath, ".go") + "_mock.go"
+					} else {
+						outputPath = filepath.Join(filepath.Dir(inputPath), outputFile)
```

Comment:
This actually does support relative paths, e.g. `../example_mock.go`, but since the mock will be generated into a different directory than the source interface, its package name and imports may be broken. This is an issue with the existing functionality too though, and fixing it in general could be very complicated. Furthermore, generating mocks into different directories from their interfaces doesn't seem all that useful anyway.